app/src/main/java/org/mozilla/fenix/perf/Performance.kt

```diff
@@ -23,27 +23,29 @@ object Performance {
     private const val EXTRA_IS_PERFORMANCE_TEST = "performancetest"
 
     /**
-     * Instruments cold startup time for use with our internal measuring system, FNPRMS. This may
-     * also appear in Google Play Vitals dashboards.
+     * Instruments "visually complete" cold startup time to homescreen for use with our internal
+     * measuring system, FNPRMS. This value may also appear in the Google Play Vitals dashboards.
      *
-     * This will need to be rewritten if any parts of the UI are changed to be displayed
-     * asynchronously.
+     * We expect this to be called when top sites is bound to its RecyclerView because we are visually
+     * complete when top sites is first drawn. The restored tabs appear at the same time so we don't
+     * need to instrument them independently.
      *
-     * In the current implementation, we only intend to instrument cold startup to the homescreen.
-     * To save implementation time, we ignore the fact that the RecyclerView draws twice if the user
-     * has tabs, collections, etc. open: the "No tabs" placeholder and a tab list. This
-     * instrumentation will only capture the "No tabs" draw.
+     * This will need to be rewritten if any parts of the UI are changed to be displayed asynchronously.
      */
-    fun instrumentColdStartupToHomescreenTime(activity: HomeActivity) {
+    fun onTopSitesBound(holder: TopSiteViewHolder) {
         // For greater accuracy, we could add an onDrawListener instead of a preDrawListener but:
         // - single use onDrawListeners are not built-in and it's non-trivial to write one
         // - the difference in timing is minimal (< 7ms on Pixel 2)
-        // - if we compare against another app using a preDrawListener, it should be comparable
+        // - if we compare against another app using a preDrawListener, as we are with Fennec, it
+        // should be comparable
         //
-        // Unfortunately, this is tightly coupled to the root view of HomeActivity's view hierarchy
-        activity.rootContainer.doOnPreDraw {
-            activity.reportFullyDrawn()
-        }
+        // We keep the code simple by calling this more than once, which we don't mind because:
+        // - reportFullyDrawn is a no-op if it's called more than once
+        // - top sites won't be bound often so redundant pre draw listeners aren't expected to have
+        // a significant perf impact
+        //
+        // Ideally we wouldn't cast to HomeActivity but we want to save implementation time.
+        holder.itemView.doOnPreDraw { (it.context as HomeActivity).reportFullyDrawn() }
```

Comment:
I think that's a good start, but looking at frames callback ( not the one I wrote but `systrace` version, I notice this would be called ~100-200 (it hovers between those two). So we actually miss that ending time. I have a few ideas on how to get that to that extra 100-200ms. Also, this changes with more top sites we have since we get the onPreDraw before any viewholders are rendered

Reply:
> I notice this would be called ~100-200

What does this mean? systrace's frame drawn time is called 100ms-200ms after our "fully drawn" is reported? If so, I think there are a few possible attributions:
- I instrumented the first frame drawn before the animation rather than at the end of the animation because I don't think we should include an animation duration within our measurement. I'm checking with Eric if this is what is desired & will add comments if so. If you were looking at the last frame drawn, this could be where the discrepancy came from.
- using `onPreDraw` rather than `onDraw`
- possible measurement overhead when using systrace or other profiling tools – not sure if it's real
- possible timestamps that don't line up between profiling tools & runtime

Keep in mind our current goal is to compare Fennec to Fenix so we only need for them to be consistent with one another, not necessarily 100% accurate (which is more complex and thus maybe not worth it right now).

Reply:
So if our goal is only to be close to Fennec, I say let's go ahead with this. Only caveat to the comment is that we're missing the onBind (so layout pass) of the view holders + the animation.  If our goal is to be close to fennec, we would need to maybe get that after the onbind pass but before the animation pass